opentelemetry-sdk/src/opentelemetry/sdk/metrics/_internal/exemplar/exemplar_reservoir.py

```diff
@@ -0,0 +1,263 @@
+        self.__trace_id: Optional[str] = None
+        self.__offered: bool = False
+
+    def offer(
+        self,
+        value: Union[int, float],
+        time_unix_nano: int,
+        attributes: Attributes,
+        ctx: Context,
+    ) -> None:
+        """Offers a measurement to be sampled."""
+        self.__value = value
+        self.__time_unix_nano = time_unix_nano
+        self.__attributes = attributes
+        span = trace.get_current_span(ctx)
+        if span != INVALID_SPAN:
+            span_context = span.get_span_context()
+            self.__span_id = span_context.span_id
+            self.__trace_id = span_context.trace_id
+
+        self.__offered = True
+
+    def collect(self, point_attributes: Attributes) -> Exemplar | None:
+        """May return an Exemplar and resets the bucket for the next sampling period."""
+        if not self.__offered:
+            return None
+
+        current_attributes = (
+            {
+                k: v
+                for k, v in self.__attributes.items()
+                if k not in point_attributes
+            }
+            if self.__attributes
+            else None
+        )
+
+        exemplar = Exemplar(
+            current_attributes,
+            self.__value,
+            self.__time_unix_nano,
+            self.__span_id,
+            self.__trace_id,
+        )
+        self.__reset()
+        return exemplar
+
+    def __reset(self) -> None:
+        self.__value = 0
+        self.__attributes = {}
+        self.__time_unix_nano = 0
+        self.__span_id = None
+        self.__trace_id = None
+        self.__offered = False
+
+
+class FixedSizeExemplarReservoirABC(ExemplarReservoir):
+    """Abstract class for a reservoir with fixed size."""
+
+    def __init__(self, size: int, **kwargs) -> None:
+        super().__init__(**kwargs)
+        self._size: int = size
+        self._reservoir_storage: list[ExemplarBucket] = [
+            ExemplarBucket() for _ in range(self._size)
+        ]
+
+    def maxSize(self) -> int:
+        """Reservoir maximal size"""
+        return self._size
+
+    def collect(self, point_attributes: Attributes) -> list[Exemplar]:
+        """Returns accumulated Exemplars and also resets the reservoir for the next
+        sampling period
+
+        Args:
+            point_attributes The attributes associated with metric point.
+
+        Returns:
+            a list of :class:`opentelemetry.sdk.metrics.exemplar.Exemplar`s. Returned
+            exemplars contain the attributes that were filtered out by the aggregator,
+            but recorded alongside the original measurement.
+        """
+        exemplars = filter(
+            lambda e: e is not None,
+            map(
+                lambda bucket: bucket.collect(point_attributes),
+                self._reservoir_storage,
+            ),
+        )
+        self._reset()
+        return [*exemplars]
+
+    def _reset(self) -> None:
+        """Reset the reservoir."""
```

Comment:
Could you clarify in the docstring here that `reset` is used for resetting any stateful logic after a collection cycle?